### PR TITLE
Documentation for code

### LINKED ISSUE #16: Detailed documentation in code
- Follow Go Docs standards and add comments to every function.

### PR DESCRIPTION
**Pull request by balandi1**
- Added comments for every function and package.
- Merging of this pull request closes Issue 16.

### CODE REVIEW

File: client/constants.js [hunk 1]
```diff
@@ -1,3 +1,7 @@
+/*
```
Comment (pradykaushik) [inline client/constants.js]:
```
// Supported interpreted languages.
const interpretedLanguages = ["python"];
// Allowed submission file types.
const allowedFiles = ["application/zip", "application/x-tar","application/gzip"];
// Allowed submission file extensions.
const allowedExtensions = [".zip", ".tar", ".tar.gz"];
```

File: client/submitAssignment.js [hunk 2]
```diff
@@ -4,7 +4,11 @@ let validCompileCmd = false;
 let supportedLanguage = "";
 let cmdArgsEnabled = false;
 
-// Gets the supported language for current environment from the server.
+/**
```
Comment (pradykaushik) [inline client/submitAssignment.js:8]: This doesn't follow the same multi-line comment style as followed in the previous comment.
Previously followed.
```
/*
comment
comment
...
*/
```

Here.
```
/*
* comment
* comment
* ...
*/
```

Follow the accepted de-factor format for multi-line comments in javascript.

File: client/submitAssignment.js [hunk 3]
```diff
@@ -20,7 +24,11 @@ window.onload = function() {
     });
 };
 
-// Uploads the form data to server.
+/**
+ * Validates assignment tarball file and sends a POST request to server
+ * for uploading the tarball and other assignment related details.
```
Comment (pradykaushik) [inline client/submitAssignment.js]: What other assignment related details are we talking about here?

File: client/submitAssignment.js [hunk 4]
```diff
@@ -20,7 +24,11 @@ window.onload = function() {
...
+ * It logs the success and error status of the operation.
```
Comment (pradykaushik) [inline client/submitAssignment.js]: we don't need to mention this.

File: client/submitAssignment.js [hunk 5]
```diff
@@ -84,7 +100,11 @@ function buildAssignment() {
     }
 }
 
-// Runs the assignment with the given command.
+/**
+ * Validates the command to run and sends POST request to server
+ * to execute or run the uploaded assignment. It logs and displays the response
```
Comment (pradykaushik) [inline client/submitAssignment.js]: we don't need to mention what the server does here. The comments should just be related to what this function does.

File: client/submitAssignment.js [hunk 6]
```diff
@@ -120,7 +143,13 @@ function resetValidations() {
     formError.innerHTML = '';
 }
 
-// Verifies the file type (should be .zip, .tar or .tar.gz)
+/**
+ * Verifies the selected file input for valid file types and extensions.
+ * Allowed file types are 'application/zip', 'application/x-tar','application/gzip'.
```
Comment (pradykaushik) [inline client/submitAssignment.js]: As allowed file types and extensions can change, having this in the comments would mean that we have to keep this in sync with updates to the constants. Instead, we can just mention "Verifies the selected file input for valid file types and extensions."

File: client/submitAssignment.js [hunk 7]
```diff
@@ -120,7 +143,13 @@ function resetValidations() {
...
+ * Displays error in 'formError' HTML element if a file is not selected within allowed types.
```
Comment (pradykaushik) [inline client/submitAssignment.js]: we don't need to mention "'formError' HTML element"

Comment (pradykaushik) [reply client/submitAssignment.js]: fix other occurrences of the same.

File: client/submitAssignment.js [hunk 8]
```diff
@@ -173,7 +211,10 @@ function validateRunCmd() {
 
 let argCount = 1;
 
-// Adds key and argument text boxes on click of "Add" button
+/**
+ * Appends a new div element with two text boxes to the 'cmdArgs' HTML element
```
Comment (pradykaushik) [inline client/submitAssignment.js]: Again the comment is too implementation specific.

File: constants/constants.go [hunk 9]
```diff
@@ -1,3 +1,4 @@
+// Package constants provide list of constant values.
```
Comment (pradykaushik) [inline constants/constants.go]: Not informative enough.

File: server/server.go [hunk 10]
```diff
@@ -1,3 +1,5 @@
+// Package server provides primitives for starting an http server and managing different
```
Comment (pradykaushik) [inline server/server.go]: what do you mean by primitives?

File: server/server.go [hunk 11]
```diff
@@ -49,7 +54,10 @@ func getSupportedLanguage(w http.ResponseWriter, r *http.Request) {
 
 }
 
-// upload parses the client request and uploads the file.
+// upload parses the client request, reads and stores the form data.
+// It uploads the assignment tarball, stores the commands to compile and run.
```
Comment (pradykaushik) [inline server/server.go]: incorrect comment. Does this method upload files?

File: server/server.go [hunk 12]
```diff
@@ -70,7 +78,8 @@ func upload(w http.ResponseWriter, r *http.Request) {
 	w.Write(response)
 }
 
-// Builds the assignment uploaded.
+// build compiles the uploaded assignment by using the command to compile
+// It sends the compilation status or message as a response to the requester.
```
Comment (pradykaushik) [inline server/server.go]: remove "to the requester" and maybe replace with "to the client" or nothing.

File: server/server.go [hunk 13]
```diff
@@ -164,6 +174,7 @@ func run(w http.ResponseWriter, r *http.Request) {
 }
 
 // navigateToWorkDir navigates to the provided working directory of the assignment.
+// It returns the current directory name(prior to navigating to working directory) and any error encountered.
```
Comment (pradykaushik) [inline server/server.go]: space required before '('

File: server/server.go [hunk 14]
```diff
@@ -245,7 +259,9 @@ func readFormData(r *http.Request) error {
 	return decompressFile(file, fileHeader, handler)
 }
 
-// decompressFile reads and stores all files from the uploaded compressed file.
+// decompressFile decompresses the assignment tarball based on the type of compression.
+// The allowed file compression types are zip, tar and tar.gz.
```
Comment (pradykaushik) [inline server/server.go]: again no need to mention file types.

File: server/server.go [hunk 15]
```diff
@@ -386,7 +404,8 @@ func listenAndServe(wg *sync.WaitGroup, port string) {
 	}
 }
 
-// StartServer starts service at given port.
+// StartServer invokes functions to start http server at given port.
```
Comment (pradykaushik) [inline server/server.go]: "StartServer sets up the different http endpoints to be used by the client and starts the http server at the given port.".

File: server/server.go (hunk 10 above)
Comment (pradykaushik) [reply server/server.go]: I see that you have used this word throughout the code base.

File: server/server_test.go [hunk 16]
```diff
@@ -16,7 +18,9 @@ import (
 
 var portNumber = flag.String("port", "52453", "Port number for server to listen on")
 
-// Creates a new file upload http request with params as cmd arguments.
+// newRequest creates a new file upload http request. This request constitutes of a tarball file,
```
Comment (pradykaushik) [inline server/server_test.go]: unnecessary comma before 'and'.

**Review by pradykaushik — CHANGES_REQUESTED** [review]

File: client/submitAssignment.js (hunk 2 above)
Comment (balandi1) [reply client/submitAssignment.js:8]: The one followed here is the format that javascript provides for JSDoc, similar to goDoc. The one previously followed is just the multi line comment.

File: client/submitAssignment.js (hunk 3 above)
Comment (balandi1) [reply client/submitAssignment.js]: No, its only the tarball. I removed the addtional comment.

File: client/submitAssignment.js (hunk 4 above)
Comment (balandi1) [reply client/submitAssignment.js]: Are you talking about the last line about logging success and error message?

File: client/submitAssignment.js (hunk 5 above)
Comment (balandi1) [reply client/submitAssignment.js]: Okay

File: client/submitAssignment.js (hunk 6 above)
Comment (balandi1) [reply client/submitAssignment.js]: Okay

File: client/submitAssignment.js (hunk 7 above)
Comment (balandi1) [reply client/submitAssignment.js]: Done

File: server/server.go (hunk 10 above)
Comment (balandi1) [reply server/server.go]: I meant code segments.

File: client/submitAssignment.js (hunk 8 above)
Comment (balandi1) [reply client/submitAssignment.js]: Okay.

File: constants/constants.go (hunk 9 above)
Comment (balandi1) [reply constants/constants.go]: Okay.

File: server/server.go (hunk 12 above)
Comment (balandi1) [reply server/server.go]: Okay

File: server/server.go (hunk 13 above)
Comment (balandi1) [reply server/server.go]: Okay

File: server/server.go (hunk 14 above)
Comment (balandi1) [reply server/server.go]: Okay

File: server/server.go (hunk 15 above)
Comment (balandi1) [reply server/server.go]: Done

File: server/server_test.go (hunk 16 above)
Comment (balandi1) [reply server/server_test.go]: Done

File: server/server.go (hunk 11 above)
Comment (balandi1) [reply server/server.go]: Agreed. Removed the upload part from the comment

File: client/constants.js (hunk 1 above)
Comment (balandi1) [reply client/constants.js]: Added

File: client/submitAssignment.js (hunk 2 above)
Comment (pradykaushik) [reply client/submitAssignment.js:8]: I don't comprehend. Why the difference in multi-line comment formats for constants.js and submitAssignment.js?

File: client/submitAssignment.js (hunk 4 above)
Comment (pradykaushik) [reply client/submitAssignment.js]: Yes.

File: constants/constants.go [hunk 17]
```diff
@@ -1,3 +1,5 @@
+// Package constants provide list of constant values
+// for the names of form elements and file type.
```
Comment (pradykaushik) [inline constants/constants.go:2]: raise an issue for adding style guides to the project to make sure that any piece of code conforms to the agreed-upon coding style and best practices.

File: server/server.go (hunk 10 above)
Comment (pradykaushik) [reply server/server.go]: Hmm. Code segments also doesn't seem apropos. I suggest taking inspiration from package comments in the standard library.

**Review by pradykaushik — CHANGES_REQUESTED** [review]

File: server/server.go (hunk 10 above)
Comment (balandi1) [reply server/server.go]: Changes made couple of days ago

File: client/submitAssignment.js (hunk 2 above)
Comment (balandi1) [reply client/submitAssignment.js:8]: Kept the multi line comment style consistent

**Review by pradykaushik — APPROVED** [review]

**Pull request merged**